### ISSUE TITLE
Panasonic no signal after manual scan

### ISSUE THREAD

**Issue by 9000h** [catalinii/minisatip#502]
there seems to be an "no signal" issue after a manual dvb scan where the TV shows "no signal" after a successful scan. 
http://paste.ubuntu.com/p/GMnyDTX64w/

**Comment by sumulu** [comment]
Maybe it has something to do with the external SAT-IP tuner? Looks like your external SATIP server is a FritzBox-Cable. I have the same problem with a Selfsat-IP server and a Panasonic. If I power on the TV, I got the channel. After switching to the next channel, I got the error. In error case, I see as PID "-1" on the external server. It's working with older version of minisatip.

**Comment by sumulu** [comment]
Same here! "no signal" or "no channel available" on Panasonic-TV with externel Selfsat-IP tuner after first switch of the channel! It's working, if I turn on the TV. External SATIP server shows in error case PID "-1". Problem starts in commit "c7f7d31f866eaee8a0a80dfa37c6158738799e9c" --> "Fix compare_slave_parameters exit code". Commit "bd80ceff5897cab46aeb2f52c30248f5ff750f0a" works for me!

**Comment by 9000h** [comment]
my test case is slightly different, the no signal issue is happen also before the commit  "c7f7d31f866eaee8a0a80dfa37c6158738799e9c" in my setup

**Comment by sumulu** [comment]
I have to say that it happen from time to time, if I switch the channels. But not so many times as before. Which version does work for you?

**Comment by 9000h** [comment]
cant remember when the issue did start, could be a early 0.7

**Comment by 9000h** [comment]
is the difference here per purpose
grep -Hn "|| sid->rtcp" stream.c 
stream.c:369:	if (sid->rtcp_sock > 0 || sid->rtcp > 0)
stream.c:482:		if (sid->rtcp_sock > 0 || sid->rtcp)

**Comment by 9000h** [comment]
https://github.com/catalinii/minisatip/commit/a6b3ebbde0c2c2fdf19f5063429c537fc2b6ddc1 could be the start of the issue ad->status_cnt is now probably not used right or dead code

**Comment by 9000h** [comment]
we may have lost "do just max 3 signal check 1s after tune"
with this patch it did work
```
--- adapter.c.orig	2018-12-04 00:43:12.545952299 +0100
+++ adapter.c	2018-12-04 00:44:40.999610675 +0100
@@ -1340,7 +1340,7 @@
 
 	if (use_ad)
 	{
-		if (ad->status < 0)
+		if (ad->status < 0 && ad->status_cnt < 8 && ad->status_cnt++ > 4)
 		{
 			status = strength = snr = 0;
 		}
```

**Comment by 9000h** [comment]
could it be that
```
[25/11 23:01:38.666 main]: Detected stream with the same destination as sid 1: sid 0 -> 192.168.178.144:3456, aid: 0
[25/11 23:01:38.666 main]: closing stream 0

````
after the second SETUP after PLAY in the same session (no TEARDOWN happen) the first stream get closed including the rtcp

**Comment by 9000h** [comment]
I did some tests and it looks like the issue did come from close_http, if I comment the exception it seems to work correct at least with VDR and/or Panasonic
```
--- minisatip.c.orig	2018-12-05 20:50:07.476488985 +0100
+++ minisatip.c	2018-12-05 20:49:44.777010452 +0100
@@ -1431,9 +1431,9 @@
 	s->buf = NULL;
 	LOG("Requested stream close %d timeout %d type %d", s->sid,
 		sid ? sid->timeout : -1, sid ? sid->type : -1);
-	if (sid && ((sid->type == STREAM_RTSP_UDP && sid->timeout != 0) || (sid->type == 0 && sid->timeout != 0)))
+//	if (sid && ((sid->type == STREAM_RTSP_UDP && sid->timeout != 0) || (sid->type == 0 && sid->timeout != 0)))
 		// Do not close rtsp udp as most likely there was no TEARDOWN at this point
-		return 0;
+//		return 0;
 	close_stream(s->sid);
 	return 0;
 }
```
what do you think will it break something else?

**Comment by 9000h** [comment]
I have a IOS SAT>IP app which is now not available atm. in the APP Store that did break, but from the logs the APP seems to be also buggy.

**Comment by catalinii** [comment]
If you comment those 2 lines, each time a UDP stream closes the TCP connection, the stream will close. This will break the satip compliance, because for a stream (udp) you are still supposed to send the stream until the timeout is reached. I think some apps relies on this way of communication with the server.

**Comment by 9000h** [comment]
right, the issue for the Panasonic is the RTCP SR which did stop after the second SETUP and switch to stream=2
![image](https://user-images.githubusercontent.com/3009073/49566540-3b622d80-f92b-11e8-9137-7c257f1ce527.png)

**Comment by catalinii** [comment]
And those 2 lines fixes the issue ?

I think the question is why rtcp (SR) stops... basically they are supposed to be send only after play (which are not).

Do u have the logs ?

**Comment by 9000h** [comment]
yes you can look at the 2 in this thread, I can also provide new once later today

**Comment by 9000h** [comment]
here the relevant part
```
[26/11 08:21:48.626 main]: SETUP rtsp://192.168.178.129:554/?src=1&freq=10744.00&pol=h&ro=0.35&msys=dvbs&mtype=qpsk&plts=off&sr=22000&fec=56 RTSP/1.0
CSeq: 1
Transport: RTP/AVP;unicast;client_port=3456-3457


[26/11 08:21:48.626 main]: detect_dvb_parameters (S)-> src=1&freq=10744.00&pol=h&ro=0.35&msys=dvbs&mtype=qpsk&plts=off&sr=22000&fec=56
[26/11 08:21:48.626 main]: detect_dvb_parameters (E) -> src=1, fe=0, freq=10744000, fec=5, sr=22000000, pol=2, ro=0, msys=5, mtype=0, plts=1, bw=8000000, inv=2, pids=NULL - apids=NULL - dpids=NULL x_pmt=NULL
[26/11 08:21:48.626 main]: Setup stream -1 parameters, sock_id 7, handle 7
[26/11 08:21:48.626 main]: stream.c:196 get_sid returns NULL for s_id = -1
[26/11 08:21:48.626 main]: Mutex init 0x560a25708088
[26/11 08:21:48.626 main]: set_sock_lock: sock_id 7 locks also mutex 0x560a2570808
````
watch TV
SETUP
PLAY stream=1
....
start channel scan on the same channel/freq working fine
SETUP
PLAY stream=2
going back in the menu where I can start the scan again and wait some seconds
sid 0 is closed due timeout (no signal)
PLAY stream=2 no RTCP

**Comment by 9000h** [comment]
did had a look at an old tcpdump from Digibit R1 / Panasonic from 2015 an a switch from stream=1 to stream=2 did also happen but the RTCP SR's did stay, so it's really only the RTCP SR missing in minisatip.

**Comment by 9000h** [comment]
int send_rtcp(int s_id, int64_t ctime)
should this not like
```
--- stream.c.orig	2018-12-06 16:15:51.609732058 +0100
+++ stream.c	2018-12-06 16:16:11.765973271 +0100
@@ -660,7 +660,7 @@
 	char ra[50];
 	unsigned char rtcp_buf[1600];
 	unsigned char *rtcp = rtcp_buf + 4;
-	streams *sid = get_sid(s_id);
+	streams *sid = get_sid_for(s_id);
 
 	if (!sid)
 		LOG_AND_RETURN(0, "Sid is null for s_id %d", s_id);
```
but it did also not fix the issue

**Comment by 9000h** [comment]
one strange thing is "Acknowledging stream %d via rtcp packet" did not show up in any logs
could this be the reason
https://github.com/catalinii/minisatip/blob/ece09b3df031a287194f2bae047169fb39fc8613/src/stream.c#L506

**Comment by 9000h** [comment]
the only reliable workaround I have so far found is
```
--- socketworks.c.orig	2018-12-15 13:47:34.460292591 +0100
+++ socketworks.c	2018-12-15 13:47:47.242028629 +0100
@@ -864,7 +864,7 @@
 						else
 							err_str = strerror(err);
 
-						if (ss->type == TYPE_RTCP || ss->sock == SOCK_TIMEOUT)
+						if (ss->type == TYPE_RTCP || ss->sock == SOCK_TIMEOUT || ss->type == TYPE_UDP)
 						{
 							LOG(
 								"ignoring error on sock_id %d handle %d type %d error %d : %s",
```
but could this issue triggered by:
```
[15/12 13:42:15.653 main]: Most likely not an RTSP packet sock_id: 16 sid: 0 rlen: 0, dropping ....
```

**Comment by 9000h** [comment]
sorry to insist  here but do you have any idea to make a proper fix?

**Comment by catalinii** [comment]
I am still trying to figure out what the issue is here:

1) if the issue is fixed by adding:
+ if (ss->type == TYPE_RTCP || ss->sock == SOCK_TIMEOUT || ss->type == TYPE_UDP)
it means that the RTP socket is being closed (for some reason). If that is the case please add logs (and point out the timestamp) when this issue happens.

2) Related to RTCP: if it is SETUP after PLAY, there are 2 cases (see https://github.com/catalinii/minisatip/blob/ece09b3df031a287194f2bae047169fb39fc8613/src/stream.c#L452 )
 - SETUP contains a new udp port in which case the old port is closed and a new one is created
 - SETUP contains the same port when the procedure above is ignored.
This case should be logged but I do not see an occurence of this.

3) The thing you pointed out in the code is basically a close from the RTSP (TCP control) socket. 
In this case as you see in:
Requested stream close 0 timeout 30000 type 2
the stream is not closed, only the socket attached to the RTSP socket, because according to the standard the stream should not be closed unless the timeout expires .

I am not sure what is going on, the only thing suspicious is that it happens multiple time close without TEARDOWN (and no idea why). 

Steps forward:
1) Lets look at this case first. Upload the logs with the patch and also upload a tcpdump of all the traffic between the 2 hosts.

**Comment by 9000h** [comment]
will do so

**Comment by 9000h** [comment]
so did the tcpdump and log with the current git
https://gofile.io/?c=mzR8FW

**Comment by catalinii** [comment]
Do you have a timestamp when the issue happened?

**Comment by 9000h** [comment]
I believe it's happen after
[20/07 15:44:29.977 AD0]: Processing SDT for transponder 1051 (41b) with length 109, sdt[5] C3

**Comment by catalinii** [comment]
Seems the tcpdump went way. Do you still have it ?

**Comment by 9000h** [comment]
no I do not have archived it but will create a new one

**Comment by 9000h** [comment]
try fix PR
part of  log plain vanilla
```
Nov 26 19:43:18 localhost minisatip[19717]: OPTIONS rtsp://192.168.178.1:554/ RTSP/1.0
Nov 26 19:43:18 localhost minisatip[19717]: OPTIONS rtsp://192.168.178.1:554/ RTSP/1.0
Nov 26 19:43:18 localhost minisatip[19717]: Public: OPTIONS, DESCRIBE, SETUP, PLAY, TEARDOWN
Nov 26 19:43:18 localhost minisatip[19717]: Public: OPTIONS, DESCRIBE, SETUP, PLAY, TEARDOWN
Nov 26 19:43:26 localhost minisatip[19717]: [26/11 19:43:26.085 main]: SETUP rtsp://192.168.178.129:554/?src=1&freq=10744.00&pol=h&ro=0.35&msys=dvbs&mtype=qpsk&plts=off&sr=22000&fec=56 RTSP/1.0
Nov 26 19:43:26 localhost minisatip[19717]: com.ses.streamID: 1
Nov 26 19:43:26 localhost minisatip[19717]: [26/11 19:43:26.091 main]: PLAY rtsp://192.168.178.129:554/stream=1?pids=0,1,2,16,17,18,20 RTSP/1.0
Nov 26 19:43:26 localhost minisatip[19717]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=8130;rtptime=961945
Nov 26 19:43:27 localhost minisatip[19717]: [26/11 19:43:27.594 main]: PLAY rtsp://192.168.178.129:554/stream=1?addpids=400 RTSP/1.0
Nov 26 19:43:27 localhost minisatip[19717]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=9436;rtptime=961946
Nov 26 19:43:27 localhost minisatip[19717]: [26/11 19:43:27.816 main]: PLAY rtsp://192.168.178.129:554/stream=1?addpids=470,1276,2171,401,402,404 RTSP/1.0
Nov 26 19:43:27 localhost minisatip[19717]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=9659;rtptime=961947
Nov 26 19:43:34 localhost minisatip[19717]: [26/11 19:43:34.912 main]: PLAY rtsp://192.168.178.129:554/stream=1?pids=0,1,16,17,18,20 RTSP/1.0
Nov 26 19:43:34 localhost minisatip[19717]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=16756;rtptime=961954
Nov 26 19:43:43 localhost minisatip[19717]: OPTIONS rtsp://192.168.178.1:554/ RTSP/1.0
Nov 26 19:43:43 localhost minisatip[19717]: OPTIONS rtsp://192.168.178.1:554/ RTSP/1.0
Nov 26 19:43:43 localhost minisatip[19717]: Public: OPTIONS, DESCRIBE, SETUP, PLAY, TEARDOWN
Nov 26 19:43:43 localhost minisatip[19717]: Public: OPTIONS, DESCRIBE, SETUP, PLAY, TEARDOWN
Nov 26 19:43:52 localhost minisatip[19717]: [26/11 19:43:52.899 main]: DESCRIBE rtsp://192.168.178.129:554/stream=1 RTSP/1.0
Nov 26 19:43:52 localhost minisatip[19717]: a=control:stream=1
Nov 26 19:43:52 localhost minisatip[19717]: [26/11 19:43:52.962 main]: SETUP rtsp://192.168.178.129:554/?src=1&freq=10744.00&pol=h&ro=0.35&msys=dvbs&mtype=qpsk&plts=off&sr=22000&fec=56 RTSP/1.0
Nov 26 19:43:52 localhost minisatip[19717]: com.ses.streamID: 2
Nov 26 19:43:52 localhost minisatip[19717]: [26/11 19:43:52.979 main]: PLAY rtsp://192.168.178.129:554/stream=2?pids=0,1,16,17,18,20 RTSP/1.0
Nov 26 19:43:53 localhost minisatip[19717]: RTP-Info: url=rtsp://192.168.178.129:554/stream=2;seq=35018;rtptime=961972
Nov 26 19:44:08 localhost minisatip[19717]: [26/11 19:44:08.950 main]: DESCRIBE rtsp://192.168.178.129:554/stream=2 RTSP/1.0
Nov 26 19:44:08 localhost minisatip[19717]: a=control:stream=2
Nov 26 19:44:09 localhost minisatip[19717]: [26/11 19:44:09.011 main]: SETUP rtsp://192.168.178.129:554/?src=1&freq=10744.00&pol=h&ro=0.35&msys=dvbs&mtype=qpsk&plts=off&sr=22000&fec=56 RTSP/1.0
Nov 26 19:44:09 localhost minisatip[19717]: com.ses.streamID: 1
Nov 26 19:44:09 localhost minisatip[19717]: [26/11 19:44:09.031 main]: PLAY rtsp://192.168.178.129:554/stream=1?pids=0,1,16,17,18,20 RTSP/1.0
Nov 26 19:44:09 localhost minisatip[19717]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=51270;rtptime=961988
Nov 26 19:44:18 localhost minisatip[19717]: [26/11 19:44:18.939 main]: PLAY rtsp://192.168.178.129:554/stream=1?pids=0,1,16,17,18,20 RTSP/1.0
Nov 26 19:44:18 localhost minisatip[19717]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=60782;rtptime=961998
Nov 26 19:44:24 localhost minisatip[19717]: [26/11 19:44:24.924 main]: PLAY rtsp://192.168.178.129:554/stream=1?pids=0,1,16,17,18,20 RTSP/1.0
Nov 26 19:44:24 localhost minisatip[19717]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=66766;rtptime=962004
Nov 26 19:44:26 localhost minisatip[19717]: [26/11 19:44:26.402 main]: PLAY rtsp://192.168.178.129:554/stream=1?pids=0,1,16,17,18,20 RTSP/1.0
Nov 26 19:44:26 localhost minisatip[19717]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=68245;rtptime=962005
Nov 26 19:44:29 localhost minisatip[19717]: [26/11 19:44:29.978 main]: TEARDOWN rtsp://192.168.178.129:554/stream=1 RTSP/1.0
```
with the PR it looks a bit better as the client did use the same stream but still did loose the signal in between 
```
Nov 26 20:20:48 localhost minisatip[7287]: [26/11 20:20:48.345 main]: SETUP rtsp://192.168.178.129:554/?src=1&freq=10744.00&pol=h&ro=0.35&msys=dvbs&mtype=qpsk&plts=off&sr=22000&fec=56 RTSP/1.0
Nov 26 20:20:48 localhost minisatip[7287]: com.ses.streamID: 1
Nov 26 20:20:48 localhost minisatip[7287]: [26/11 20:20:48.351 main]: PLAY rtsp://192.168.178.129:554/stream=1?pids=0,1,16,17,18,20 RTSP/1.0
Nov 26 20:20:48 localhost minisatip[7287]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=881673;rtptime=964187
Nov 26 20:20:49 localhost minisatip[7287]: [26/11 20:20:49.817 main]: PLAY rtsp://192.168.178.129:554/stream=1?addpids=400 RTSP/1.0
Nov 26 20:20:49 localhost minisatip[7287]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=882931;rtptime=964189
Nov 26 20:20:49 localhost minisatip[7287]: [26/11 20:20:49.933 main]: PLAY rtsp://192.168.178.129:554/stream=1?addpids=470,1276,2171,401,402,404 RTSP/1.0
Nov 26 20:20:49 localhost minisatip[7287]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=883049;rtptime=964189
Nov 26 20:20:58 localhost minisatip[7287]: [26/11 20:20:58.602 main]: PLAY rtsp://192.168.178.129:554/stream=1?pids=0,1,16,17,18,20 RTSP/1.0
Nov 26 20:20:58 localhost minisatip[7287]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=891718;rtptime=964197
Nov 26 20:21:13 localhost minisatip[7287]: [26/11 20:21:13.478 main]: DESCRIBE rtsp://192.168.178.129:554/stream=1 RTSP/1.0
Nov 26 20:21:13 localhost minisatip[7287]: a=control:stream=1
Nov 26 20:21:13 localhost minisatip[7287]: [26/11 20:21:13.535 main]: SETUP rtsp://192.168.178.129:554/?src=1&freq=10744.00&pol=h&ro=0.35&msys=dvbs&mtype=qpsk&plts=off&sr=22000&fec=56 RTSP/1.0
Nov 26 20:21:13 localhost minisatip[7287]: com.ses.streamID: 1
Nov 26 20:21:13 localhost minisatip[7287]: [26/11 20:21:13.546 main]: PLAY rtsp://192.168.178.129:554/stream=1?pids=0,1,16,17,18,20 RTSP/1.0
Nov 26 20:21:13 localhost minisatip[7287]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=906901;rtptime=964213
Nov 26 20:21:17 localhost minisatip[7287]: [26/11 20:21:17.311 main]: PLAY rtsp://192.168.178.129:554/stream=1?pids=0,1,16,17,18,20 RTSP/1.0
Nov 26 20:21:17 localhost minisatip[7287]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=910426;rtptime=964216
Nov 26 20:21:26 localhost minisatip[7287]: [26/11 20:21:26.200 main]: PLAY rtsp://192.168.178.129:554/stream=1?pids=0,1,16,17,18,20 RTSP/1.0
Nov 26 20:21:26 localhost minisatip[7287]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=919315;rtptime=964225
Nov 26 20:21:27 localhost minisatip[7287]: [26/11 20:21:27.815 main]: PLAY rtsp://192.168.178.129:554/stream=1?pids=0,1,16,17,18,20 RTSP/1.0
Nov 26 20:21:27 localhost minisatip[7287]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=920929;rtptime=964227
Nov 26 20:21:30 localhost minisatip[7287]: [26/11 20:21:30.350 main]: PLAY rtsp://192.168.178.129:554/stream=1?pids=0,1,16,17,18,20 RTSP/1.0
Nov 26 20:21:30 localhost minisatip[7287]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=923464;rtptime=964229
Nov 26 20:21:31 localhost minisatip[7287]: [26/11 20:21:31.888 main]: PLAY rtsp://192.168.178.129:554/stream=1?pids=0,1,16,17,18,20 RTSP/1.0
Nov 26 20:21:31 localhost minisatip[7287]: RTP-Info: url=rtsp://192.168.178.129:554/stream=1;seq=925002;rtptime=964231
Nov 26 20:21:35 localhost minisatip[7287]: [26/11 20:21:35.764 main]: TEARDOWN rtsp://192.168.178.129:554/stream=1 RTSP/1.0

```

**Comment by 9000h** [comment]
maybe there is another bug in the Panasonic message flow

DESCRIBE rtsp://192.168.178.129:554/stream=1 RTSP/1.0
this should look different at least for unicast, see 4.2.27.2 Sequence
http://www.satip.info/sites/satip/files/resource/satip_rtsp_message_flows_version_1_0_0.pdf
should come from the client like
DESCRIBE rtsp://192.168.178.129:554/ RTSP/1.0
so we may need to ignore the stream=1 here

**Comment by 9000h** [comment]
the strange packets 144 is panasonic  129 minisatip
22222	2019/332 18:02:02,540775	192.168.178.144	192.168.178.129	ICMP	72	Destination unreachable (Port unreachable)
new log and dump
[dump-pana.zip](https://github.com/catalinii/minisatip/files/3904928/dump-pana.zip)

[dump-pana.log](https://github.com/catalinii/minisatip/files/3904935/dump-pana.log)

**Comment by lars18th** [comment]
> maybe there is another bug in the Panasonic message flow
> 
> DESCRIBE rtsp://192.168.178.129:554/**stream=1** RTSP/1.0
> this should look different at least for unicast, see 4.2.27.2 Sequence
> http://www.satip.info/sites/satip/files/resource/satip_rtsp_message_flows_version_1_0_0.pdf
> should come from the client like
> DESCRIBE rtsp://192.168.178.129:554/ RTSP/1.0
> so we may need to ignore the stream=1 here or all behind the /
> the answer from minisatip contains **Session:** which should not given to the client
> 
> Nov 28 09:36:35 localhost minisatip[32011]: Date: Thu, Nov 28 08:36:35 2019 GMT
> **Nov 28 09:36:35 localhost minisatip[32011]: Session: 1804289383**
> Nov 28 09:36:35 localhost minisatip[32011]: CSeq: 2
> Nov 28 09:36:35 localhost minisatip[32011]: Server: minisatip/1.0.2-f4a3400
> Nov 28 09:36:35 localhost minisatip[32011]: Content-type: application/sdp
> Nov 28 09:36:35 localhost minisatip[32011]: Content-Base: rtsp://192.168.178.129/
> Nov 28 09:36:35 localhost minisatip[32011]: Content-Length: 250

This issue is fixed in my PR #657 .

**Comment by lars18th** [comment]
> there are also strange icmp packets (144 is panasonic 129 minisatip)
> 
> 22222 2019/332 18:02:02,540775 192.168.178.144 192.168.178.129 ICMP 72 Destination unreachable (Port unreachable)

I see this type of errors when the CLIENT stops to read a RTP stream. After it starts to read, the TCP/IP  stack is aware of the flow. So when the flow stops, the SERVER TCP/IP stack can receive this message to understand that the client refuses the flow.

But regarding the problem with the Panasonic TV... I feel the problem is that the minisatip server changes his internal state.

**Comment by lars18th** [comment]
Hi @9000h ,

To get the data between the Panasonic TV and the minisatip without the wireshark, you can use another alternative:

- Compile in the same machine when you run the `minisatip` this fork of `sat-ip-proxy`: https://github.com/lars18th/sat-ip-proxy
- Then execute the `minisatip` in an alternative RTSP port, for example: `minisatip -y 5544`.
- Execute the proxy with `./rtsp-proxy -dd 192.168.178.129:5544` (use the LAN ip of the minisatip sever).
- Then power-on your TV and see the messages.

I hope it helps. :wink:

**Comment by 9000h** [comment]
I have the impression at least 2 things in combination ar happen
```
Nov 28 18:02:02 localhost minisatip[17358]: a=sendonly
Nov 28 18:02:02 localhost minisatip[17358]: [28/11 18:02:02.529 main]: read RTSP (from handle 24 sock 16, len: 0, sid 0):
Nov 28 18:02:02 localhost minisatip[17358]: [28/11 18:02:02.529 main]:
Nov 28 18:02:02 localhost minisatip[17358]: [28/11 18:02:02.529 main]: Most likely not an RTSP packet sock_id: 16 sid: 0 rlen: 0, dropping ....
Nov 28 18:02:02 localhost minisatip[17358]: [28/11 18:02:02.529 main]: select_and_execute[16]: Close on socket 24 (sid:0) from 192.168.178.144:26202 - type rtsp errno 0
Nov 28 18:02:02 localhost minisatip[17358]: [28/11 18:02:02.529 main]: Requested stream close 0 timeout 30000 type 2
```
where the socket get closed and the RTCP get in timeout and stops
and a new stream get assigned a the old is already -1
https://github.com/catalinii/minisatip/blob/f4a3400980a803e014ed29deee451121f47f7e70/src/stream.c#L95-L98

**Comment by lars18th** [comment]
> * Execute the proxy with `./rtsp-proxy -dd 192.168.178.129:5544` (use the LAN ip of the minisatip sever).

- `-dd` to see only the RTSP messages.
- `-ddd` to see the RTCP messages too.
- `-dddd` to see all messages (including RTP).

**Comment by lars18th** [comment]
Hi @9000h ,

> I have the impression at least 2 things in combination ar happen
> 
> [...] Most likely not an RTSP packet sock_id: 16 sid: 0 rlen: 0, dropping ....
>
> where the socket get closed and the RTCP get in timeout and stops
> and a new stream get assigned a the old is already -1

Note here the `rlen: 0`. The RTSP socket has received 0 bytes.
I feel this is related to: https://github.com/catalinii/minisatip/issues/611#issuecomment-539363212

Please, use the proxy and check if the RTSP session is closed by the Panasonic or by the minisatip server.

**Comment by 9000h** [comment]
IMHO it should not matter if the client did close,  zero length RTSP should not be processed by rtsp_read() it is one trigger of the issue

**Comment by lars18th** [comment]
> IMHO it should not matter if the client did close, zero length RTSP should not be processed by rtsp_read() it is one trigger of the issue

Yes. But we need to isolate the cause first.

**Comment by 9000h** [comment]
EAGAIN is may equal to EWOULDBLOCK is maybe also worse a look

With glibc, those two are equal, but may not with other libraries.

in minisatip socketworks.c it's used and may not work as expected depending on the system libs

it may should be handled like:

#if EAGAIN == EWOULDBLOCK
  #define IS_EAGAIN(e) (e == EAGAIN)
#else
  #define IS_EAGAIN(e) (e == EAGAIN || e == EWOULDBLOCK)
#endif

**Comment by lars18th** [comment]
Hi @9000h ,

I do a review of your pcap file from today... and I don't see any error! The RTSP session seems correct.

Please, can you describe _another time_ the problem with your Panasonic TV?

**Comment by 9000h** [comment]
sure it look good, but the minisatip log shows the dopping message.

the rlen=0 in rtsp is also happen with a VDR client and seems to be general issue,

for the Panasonic it's maybe only a part of the issue.

**Comment by lars18th** [comment]
Hi @9000h ,

> sure it look good, but the minisatip log shows the dopping message.
> 
> the rlen=0 in rtsp is also happen with a VDR client and seems to be general issue,
> 
> for the Panasonic it's maybe only a part of the issue.

I need to comment that with my new (2019) Panasonic TV the current minisatip works well. And my troubles are mainly with the `satipc` implementation (at time minisatip doesn't like all SAT>IP servers). Futhermore, I don't use VDR. So the question is more simple: which erroneous behaviour you see with _your_ Panasonic TV ?

**Comment by lars18th** [comment]
Hi @9000h ,

I forget to mention that in my environment (a slow minisatip server, low cost STB as the SAT>IP server and a 2019 Panasonic TV) I need to use this: `minisatip -i 32 -H 1:15 -b 1316:5775360`. This fixes a lot of problems with the _SMALL_ buffer that the Panasonic's implementation has.

Regards.

**Comment by lars18th** [comment]
> the rlen=0 in rtsp is also happen with a VDR client and seems to be general issue,

I'm not sure, but I suspect that something inside the current code is closing some TCP sockets randomly.

**Comment by 9000h** [comment]
I do not have general issues with the Panasonic's and minisatip, but when you do a manual dvb scan repeated it will loose the signal, and when you go out of the manual scan the signal did not come back.
In this case I have to switch to DVB-C/S and then back to SAT>IP to recover the signal.

**Comment by 9000h** [comment]
so simple test on Ubuntu 19.04 intel
```
#include <stdio.h>
#include <netinet/in.h>
void main()
{

#if EAGAIN == EWOULDBLOCK
#define IS_EAGAIN(e) (e == EAGAIN)
 printf("EAGAIN == EWOULDBLOCK \n");
#else
#define IS_EAGAIN(e) (e == EAGAIN || e == EWOULDBLOCK)
 printf("EAGAIN != EWOULDBLOCK \n");
#endif

}
root@ubuntu:/tmp# cc t.c -o t
root@ubuntu:/tmp# ./t 
EAGAIN == EWOULDBLOCK 
````
this proves we may a issue here at least with glibc
https://github.com/catalinii/minisatip/blob/f4a3400980a803e014ed29deee451121f47f7e70/src/socketworks.c#L897-L903

**Comment by lars18th** [comment]
> this proves we may a issue here at least with glibc
> `EAGAIN == EWOULDBLOCK`

This proves it too: https://stackoverflow.com/questions/49049430/difference-between-eagain-or-ewouldblock

I feel the code requires then to be:
```
 if (err == EOVERFLOW) 
 	continue; 
 if (err == EAGAIN) 
 { 
 	ss->err++; 
 	if (ss->err < 10) 
 		continue; 
```

**Comment by 9000h** [comment]
right but it's still not a 100% solution

**Comment by 9000h** [comment]
new log from git today 
[log-pana.log](https://github.com/catalinii/minisatip/files/3910088/log-pana.log)

**Comment by 9000h** [comment]
new log git today
[log-pana.log](https://github.com/catalinii/minisatip/files/3912366/log-pana.log)

**Comment by 9000h** [comment]
new log git today
[log-04122019.log](https://github.com/catalinii/minisatip/files/3919134/log-04122019.log)
so with the latest commits we getting closer
this is may the point
Dec  4 00:01:39 localhost minisatip[2391]: [04/12 00:01:39.447 main]: Detected stream with the same destination as sid 1: sid 0 -> 192.168.178.144:3456, aid: 0

**Comment by lars18th** [comment]
Hi @9000h ,

> new log git today
> [log-04122019.log](https://github.com/catalinii/minisatip/files/3919134/log-04122019.log)
> so with the latest commits we getting closer
> this is may the point
> Dec 4 00:01:39 localhost minisatip[2391]: [04/12 00:01:39.447 main]: Detected stream with the same destination as sid 1: sid 0 -> 192.168.178.144:3456, aid: 0

Regarding your log I see several issues. However, I feel I've identified your remaining problem:

- The SAT>IP Client (in this case the Panasonic TV doing the scan) sends a _new_ SETUP command _without_ a previous TEARDOWN. See this part of the log:

```
[...]
Dec  4 00:01:14 localhost minisatip[2391]: [04/12 00:01:14.453 main]: SETUP rtsp://192.168.178.129:554/?src=1&freq=10744.00&pol=h&ro=0.35&msys=dvbs&mtype=qpsk&plts=off&sr=22000&fec=56 RTSP/1.0
Dec  4 00:01:14 localhost minisatip[2391]: CSeq: 1
Dec  4 00:01:14 localhost minisatip[2391]: Transport: RTP/AVP;unicast;client_port=3456-3457
Dec  4 00:01:14 localhost minisatip[2391]: #015
[...]
Dec  4 00:01:39 localhost minisatip[2391]: [04/12 00:01:39.446 main]: SETUP rtsp://192.168.178.129:554/?src=1&freq=10744.00&pol=h&ro=0.35&msys=dvbs&mtype=qpsk&plts=off&sr=22000&fec=56 RTSP/1.0
Dec  4 00:01:39 localhost minisatip[2391]: CSeq: 1
Dec  4 00:01:39 localhost minisatip[2391]: Transport: RTP/AVP;unicast;client_port=3456-3457
Dec  4 00:01:39 localhost minisatip[2391]: #015
[...]
```

And this isn't really valid. Why? Because the client _can_ open a new session. That isn't forbiden, as the client can open as many session as it wants. However, for a _secondary_ session the target client_port requires to be different (as it's for _another_ stream). And that's the cause of the message `Detected stream with the same destination `.

Solutions to this problem:

1. The SAT>IP client sends a TEARDOWN before, or changes the SETUP with a PLAY command with new tunning parameters and the stream id. However, this is out-of-scope for us!
2. Do a trick in the minisatip to process such case. Read more...

I propose two different solutions (and I ask to @catalinii for the best solution from his point of view):
- When a new SETUP command is received, check the target client_port pair. If some of them are in use for another session then close the session (as if a TEARDOWN is received). I feel this is the most direct and simple solution to implement.
- Do the same, but instead of closing the session do a rewrite (internally) of the SETUP command with a PLAY command, adding the stream id from the detected in use session with the same target client_port. I'm sure this will work, and doesn't disturb the client. So it will be transparent (the response message for the PLAY will be identical to the expected one for the SETUP). Perhaps this will be more tricky, but I feel it will be more robust and will fix the problem.

Feel free to implement whatever you prefer.
Regards.

**Comment by lars18th** [comment]
Hi @9000h ,

Another comment about your log: I see that you doesn't have included my PR #657. Please, try it! I created it specially for you. :wink:

**Comment by lars18th** [comment]
Hi @catalinii ,

In the last log from @9000h I see a potential bug. Please, see this part of the log:
```
Dec  4 00:01:22 localhost minisatip[2391]: [04/12 00:01:22.749 main]: PLAY rtsp://192.168.178.129:554/stream=1?pids=0,1,16,17,18,20 RTSP/1.0
Dec  4 00:01:22 localhost minisatip[2391]: CSeq: 5
Dec  4 00:01:22 localhost minisatip[2391]: Session: 0846930886
Dec  4 00:01:22 localhost minisatip[2391]: #015
Dec  4 00:01:22 localhost minisatip[2391]: [04/12 00:01:22.749 main]: detect_dvb_parameters (S)-> pids=0,1,16,17,18,20
Dec  4 00:01:22 localhost minisatip[2391]: [04/12 00:01:22.749 main]: detect_dvb_parameters (E) -> src=-1, fe=-1, freq=-1, fec=-1, sr=-1, pol=-1, ro=-1, msys=-1, mtype=-1, plts=-1, bw=-1, inv=-1, pids=0,1,16,17,18,20 - apids=NULL - dpids=NULL x_pmt=NULL
Dec  4 00:01:22 localhost minisatip[2391]: [04/12 00:01:22.749 main]: Setup stream 0 parameters, sock_id 7, handle 7
Dec  4 00:01:22 localhost minisatip[2391]: [04/12 00:01:22.749 main]: copy_dvb_param start -> src=1, fe=0, freq=10744000, fec=5, sr=22000000, pol=2, ro=0, msys=5, mtype=0, plts=1, bw=8000000, inv=2, pids=NULL, apids=NULL, dpids=NULL x_pmt=NULL
Dec  4 00:01:22 localhost minisatip[2391]: [04/12 00:01:22.749 main]: copy_dvb_parameters -> src=1, fe=0, freq=10744000, fec=5 sr=22000000, pol=2, ro=0, msys=5, mtype=0, plts=1, bw=8000000, inv=2, pids=0,1,16,17,18,20, apids=NULL, dpids=NULL x_pmt=NULL
Dec  4 00:01:22 localhost minisatip[2391]: [04/12 00:01:22.749 main]: Updating stream 0 time to 60007, current time 60007
Dec  4 00:01:22 localhost minisatip[2391]: [04/12 00:01:22.750 main]: Play for stream 0, type 2, rsock 8, adapter 0, sock_id 7, rsock_id 9, handle 7
Dec  4 00:01:22 localhost minisatip[2391]: [04/12 00:01:22.750 main]: setting DVB parameters for adapter 0 - master_sid 0 sid 0 old f:10744000
Dec  4 00:01:22 localhost minisatip[2391]: [04/12 00:01:22.750 main]: copy_dvb_param start -> src=1, fe=0, freq=10744000, fec=5, sr=22000000, pol=2, ro=0, msys=5, mtype=0, plts=1, bw=8000000, inv=2, pids=NULL, apids=470, dpids=NULL x_pmt=NULL
Dec  4 00:01:22 localhost minisatip[2391]: [04/12 00:01:22.750 main]: copy_dvb_parameters -> src=1, fe=0, freq=10744000, fec=5 sr=22000000, pol=2, ro=0, msys=5, mtype=0, plts=1, bw=8000000, inv=2, pids=0,1,16,17,18,20, apids=NULL, dpids=NULL x_pmt=NULL
Dec  4 00:01:22 localhost minisatip[2391]: [04/12 00:01:22.750 main]: deleting pids on adapter 0, sid 0, pids=NULL
```

The PLAY command here uses `pids=` _after_ a previoius PLAY with `addpids=`. So the server requires to free the pids list _before_ adding the new pids. However, I see quite annoying this line of the log:
```
copy_dvb_param start -> src=1, fe=0, freq=10744000, fec=5, sr=22000000, pol=2, ro=0, msys=5, mtype=0, plts=1, bw=8000000, inv=2, pids=NULL, apids=470, dpids=NULL x_pmt=NULL
```

Why is a `apids=470` here?

**Comment by 9000h** [comment]
the apids=470 could be come from a "of by one"

**Comment by lars18th** [comment]
Hi @9000h ,

>> Why is a apids=470 here?

> the apids=470 could be come from a "of by one"

The problems with this log message are:
- This pid is _alone_ inside the list of pids to add in the _current_ adapter _before_ apply the new pid list.
- Futhermore this pid is already included in the list of _running_ pids.

So, who has requested to _add_ this pid, and only this, when it's already included? And why it's added to the _current_ adapter just bofere processing the rquest of a new pid list?

I noticed that the pid 470 is the AIT (HbbTV) of the "arte" service in the transponder 10744 of Astra 19.2º. So it isn't a PMT of some service that the minisatip wants to parse.

Any idea of this?

**Comment by 9000h** [comment]
I have also no idea where it comes from :-(

**Comment by lars18th** [comment]
> I have also no idea where it comes from :-(

I hope @catalinii will discover the problem!

**Comment by 9000h** [comment]
ther is another strange line in the log
Dec  4 00:01:22 localhost minisatip[2391]: [04/12 00:01:22.751 main]: pid 600, fd 27, packets 29, d/c errs 0/0, flags 3, pmt -4, filter 6, sock -1, sids: -1 -1 -1 -1 -1 -1 -1 -1

how can the pmt -4 (I also have seen -29 in older logs)

**Comment by lars18th** [comment]
> how can the pmt -4 (I also have seen -29 in older logs)

Yes, and `pmt -2` in the same log!

**Comment by 9000h** [comment]
the rtsp drops are NOP's len 0 packet coming from the client, any idea how we can drop them before processing in read_rtsp?
![2019-12-06 13_30_35-Window](https://user-images.githubusercontent.com/3009073/70323406-f833e680-182c-11ea-9a53-e0bd0966d7df.png)

**Comment by 9000h** [comment]
@catalinii 
as this packets are from client close what do you think about
````
--- socketworks.c       2019-12-06 14:10:17.909481119 +0100
+++ socketworks.c.new   2019-12-06 14:10:07.337653748 +0100
@@ -880,7 +880,7 @@
                                                read_ok ? "OK" : "NOK", rlen, master->rlen, master->lbuf,
                                                ss->sock, ss->id, master->id, pos, master->buf, ss->iteration, master->action);

-                                       if (((master->rlen > 0) || err == EWOULDBLOCK) && master->action && (master->type != TYPE_SERVER))
+                                       if (((master->rlen > 0) || err == EWOULDBLOCK) && master->action && (master->type != TYPE_SERVER) && err != ECONNRESET)
                                                master->action(master);

                                        sockets_unlock(ss);
````

**Comment by 9000h** [comment]
maybe here is a way as the second SETUP did also have CSeq 1

**Comment by catalinii** [comment]
Can you try to reject the new stream if it comes with the same UDP destination?
(add return -1 before this line: https://github.com/catalinii/minisatip/blob/master/src/stream.c#L527 ) 

Re: pid 470:
[04/12 00:01:22.749 main]: copy_dvb_param start -> src=1, fe=0, freq=10744000, fec=5, sr=22000000, pol=2, ro=0, msys=5, mtype=0, plts=1, bw=8000000, inv=2, pids=NULL, apids=NULL, dpids=NULL x_pmt=NULL
Dec  4 00:01:22 localhost minisatip[2391]: [04/12 00:01:22.749 main]: copy_dvb_parameters -> src=1, fe=0, freq=10744000, fec=5 sr=22000000, pol=2, ro=0, msys=5, mtype=0, plts=1, bw=8000000, inv=2, pids=0,1,16,17,18,20, apids=NULL, dpids=NULL x_pmt=NULL

This comes from the previous values. As you can see afterwards the new values in the second line are populated correctly. You should see also below that the right pids are added to the adapter.

**Comment by 9000h** [comment]
> Can you try to reject the new stream if it comes with the same UDP destination?
> (add return -1 before this line: https://github.com/catalinii/minisatip/blob/master/src/stream.c#L527 )

will try, but I believe I have done this before ending up in a loop.

**Comment by 9000h** [comment]
I did fix it already without notice, funny :-)

this https://github.com/catalinii/minisatip/commit/09adcc7c56ce2c701d16876b52c14fa09def7858 commit was the real fix!
so the Panasonic TV's clients really use the app string and expect the right order.
thanks for all your efforts!
